edge-agent/test/Microsoft.Azure.Devices.Edge.Agent.DiagnosticsComponent.Test/MerticsSerializerTests.cs

```diff
@@ -0,0 +1,113 @@
+    using System.Linq;
+    using System.Runtime.InteropServices;
+    using System.Text;
+    using Microsoft.Azure.Devices.Edge.Util.Test.Common;
+    using Xunit;
+
+    public class MerticsSerializerTests
+    {
+        readonly Random rand = new Random();
+
+        [Fact]
+        public void SingleMetricSerializes()
+        {
+            Metric testMetric = new Metric(DateTime.UtcNow, this.metrics[0].name, this.rand.NextDouble() * 50, this.metrics[0].tags);
+
+            byte[] data = RawMetric.MetricsToBytes(new Metric[] { testMetric }).ToArray();
+            var reconstructedValues = RawMetric.BytesToMetrics(data).ToArray();
+
+            Assert.Single(reconstructedValues);
+            TestUtilities.ReflectionEqual(testMetric, reconstructedValues[0]);
+        }
+
+        [Fact]
+        public void SingleMetricSeriesSerializes()
+        {
+            var testMetrics = this.GenerateSeries(this.metrics[0].name, this.metrics[0].tags).ToArray();
+
+            byte[] data = RawMetric.MetricsToBytes(testMetrics).ToArray();
+            var reconstructedValues = RawMetric.BytesToMetrics(data).ToArray();
+
+            var expected = testMetrics.OrderBy(m => m.Name).ThenBy(m => m.Tags).ThenBy(m => m.TimeGeneratedUtc);
+            var actual = reconstructedValues.OrderBy(m => m.Name).ThenBy(m => m.Tags).ThenBy(m => m.TimeGeneratedUtc);
+            TestUtilities.ReflectionEqualEnumerable(expected, actual);
+        }
+
+        [Fact]
+        public void MetricsSeriesSerializes()
+        {
+            var testMetrics = this.metrics.SelectMany(m => this.GenerateSeries(m.name, m.tags, this.rand.Next(10, 20))).ToArray();
+
+            byte[] data = RawMetric.MetricsToBytes(testMetrics).ToArray();
+            var reconstructedValues = RawMetric.BytesToMetrics(data).ToArray();
+
+            var expected = testMetrics.OrderBy(m => m.Name).ThenBy(m => m.Tags).ThenBy(m => m.TimeGeneratedUtc).ToArray();
+            var actual = reconstructedValues.OrderBy(m => m.Name).ThenBy(m => m.Tags).ThenBy(m => m.TimeGeneratedUtc).ToArray();
+            TestUtilities.ReflectionEqualEnumerable(expected, actual);
+        }
+
+        [Fact]
+        public void RawValuesSerialize()
+        {
+            var time = DateTime.UtcNow;
+            var testValues = Enumerable.Range(1, 10).Select(i => new RawMetricValue(time.AddHours(i), this.rand.NextDouble() * 200)).ToArray();
+
+            byte[] data = RawMetricValue.RawValuesToBytes(testValues).ToArray();
+            var reconstructedValues = RawMetricValue.BytesToRawValues(data, 0, testValues.Length).ToArray();
+
+            TestUtilities.ReflectionEqualEnumerable(testValues, reconstructedValues);
+        }
+
+        [Fact]
+        public void InvalidDataThrows()
+        {
+            // Gibberish
+            byte[] randData = new byte[300];
+            this.rand.NextBytes(randData);
+            Assert.Throws<InvalidDataException>(() => RawMetric.BytesToMetrics(randData).ToArray());
```

Comment:
Please assert inner exception type.

Reply:
asserted inner exception type for overflow. The inner exception for the random data is unknown. It could be any of the reasons.